olp-cpp-sdk-dataservice-read/src/StreamLayerClientImpl.cpp

```diff
@@ -20,24 +20,133 @@
 #include "StreamLayerClientImpl.h"
 
 #include <olp/core/cache/DefaultCache.h>
+#include <olp/core/client/CancellationContext.h>
 #include <olp/core/client/OlpClientSettingsFactory.h>
+#include <olp/core/logging/Log.h>
+#include <olp/core/thread/TaskScheduler.h>
+
+#include "ApiClientLookup.h"
+#include "Common.h"
+#include "generated/api/StreamApi.h"
+#include "repositories/ExecuteOrSchedule.inl"
 
 namespace olp {
 namespace dataservice {
 namespace read {
 
+namespace {
+constexpr auto kLogTag = "StreamLayerClientImpl";
+constexpr auto kSubscriptionModeParallel = "parallel";
+constexpr auto kSubscriptionModeSerial = "serial";
+constexpr auto kStreamService = "stream";
+constexpr auto kStreamVersion = "v2";
+
+std::string GetSubscriptionMode(SubscribeRequest::SubscriptionMode mode) {
+  return mode == SubscribeRequest::SubscriptionMode::kSerial
+             ? kSubscriptionModeSerial
+             : kSubscriptionModeParallel;
+}
+
+}  // namespace
+
 StreamLayerClientImpl::StreamLayerClientImpl(client::HRN catalog,
                                              std::string layer_id,
                                              client::OlpClientSettings settings)
     : catalog_(std::move(catalog)),
       layer_id_(std::move(layer_id)),
-      settings_(std::move(settings)) {
+      settings_(std::move(settings)),
+      pending_requests_(std::make_shared<client::PendingRequests>()) {
   if (!settings_.cache) {
     settings_.cache = client::OlpClientSettingsFactory::CreateDefaultCache({});
   }
 }
 
-StreamLayerClientImpl::~StreamLayerClientImpl() = default;
+StreamLayerClientImpl::~StreamLayerClientImpl() {
+  pending_requests_->CancelAllAndWait();
+}
+
+bool StreamLayerClientImpl::CancelPendingRequests() {
+  OLP_SDK_LOG_TRACE(kLogTag, "CancelPendingRequests");
+  return pending_requests_->CancelAll();
+}
+
+client::CancellationToken StreamLayerClientImpl::Subscribe(
+    SubscribeRequest request, SubscribeResponseCallback callback) {
+  using SubscribeResponse =
```

Comment:
Doesn't make much sense.
Why don't you return simply SubscribeResponse and set the member client_context_ inside the subscribe_task while the AddTask callback parameter only returns response to the user?